File: kotlin/src/main/kotlin/solutions/Day07.kt
package solutions

class Day07: Day {
    override fun part1Solution(input: List<String>): String {
        val wagers = mutableListOf<Pair<Hand, Int>>()
        input.forEach { round ->
            val roundParts = round.split(" ")
            val cards = roundParts[0].map { CamelCard.create(it) }
            wagers.add(Hand(cards, HandType.fromCards(cards)) to roundParts[1].toInt() )
        }
        wagers.sortWith(
            compareBy <Pair<Hand, Int>> {it.first.type}
                .thenBy { it.first.cards[0] }
                .thenBy { it.first.cards[1] }
                .thenBy { it.first.cards[2] }
                .thenBy { it.first.cards[3] }
                .thenBy { it.first.cards[4] }
        )
        var answer = 0
        wagers.forEachIndexed { index, pair ->
            answer += (wagers.size - index) * pair.second
        }
        return answer.toString()
    }

    override fun part2Solution(input: List<String>): String {
        val wagers = mutableListOf<Pair<Hand, Int>>()
        input.forEach { round ->
            val roundParts = round.split(" ")
            val cards = roundParts[0].map { CamelCard.create(it, true) }
            wagers.add(Hand(cards, HandType.fromCards(cards, true)) to roundParts[1].toInt() )
        }
        wagers.sortWith(
            compareBy <Pair<Hand, Int>> {it.first.type}
                .thenBy { it.first.cards[0] }
                .thenBy { it.first.cards[1] }
                .thenBy { it.first.cards[2] }
                .thenBy { it.first.cards[3] }
                .thenBy { it.first.cards[4] }
        )
        var answer = 0
        wagers.forEachIndexed { index, pair ->
            answer += (wagers.size - index) * pair.second
        }
        return answer.toString()
    }

    data class Hand(val cards: List<CamelCard>, val type: HandType)

    enum class CamelCard {
        A, K, Q, J, T, NINE, EIGHT, SEVEN, SIX, FIVE, FOUR, THREE, TWO, JOKER;

        companion object {
            fun create(char: Char, isPartTwo: Boolean = false) = when(char){
                'A' -> A
                'K' -> K
                'Q'-> Q
                'J'-> if (isPartTwo) JOKER else J
                'T'-> T
                '9'-> NINE
                '8'-> EIGHT
                '7'-> SEVEN
                '6'-> SIX
                '5'-> FIVE
                '4'-> FOUR
                '3'-> THREE
                else -> TWO
            }
        }

    }

    enum class HandType {
        FIVE_OF_A_KIND, FOUR_OF_A_KIND, FULL_HOUSE, THREE_OF_A_KIND, TWO_PAIR, ONE_PAIR, HIGH_CARD;

        companion object {
            fun fromCards(cards: List<CamelCard>, isPartTwo: Boolean = false): HandType {
                val counts = mutableMapOf<CamelCard, Int>()
                cards.forEach { card ->
                    counts[card] = counts.getOrDefault(card, 0) + 1
                }
                return if (!isPartTwo) {
                    when (counts.values.max()) {
                        5 -> FIVE_OF_A_KIND
                        4 -> FOUR_OF_A_KIND
                        3 -> if (counts.keys.size == 2) FULL_HOUSE else THREE_OF_A_KIND
                        2 -> if (counts.keys.size == 3) TWO_PAIR  else ONE_PAIR
                        else -> HIGH_CARD
                    }
                } else {
                    val maxDuplicates = counts.filter { it.key != CamelCard.JOKER }.values.maxOrNull()
                    val numJokers = counts[CamelCard.JOKER]
                    when(maxDuplicates){
                        5 -> FIVE_OF_A_KIND
                        4 -> if (numJokers == 1) FIVE_OF_A_KIND else FOUR_OF_A_KIND
                        3 -> if (numJokers == 2) {
                                FIVE_OF_A_KIND
                            } else if (numJokers == 1) {
                                FOUR_OF_A_KIND
                            } else if (counts.keys.size == 2) {
                                FULL_HOUSE
                            } else {
                                THREE_OF_A_KIND
                            }
                        2 -> if (numJokers == 3) {
                            FIVE_OF_A_KIND
                            } else if (numJokers == 2) {
                                FOUR_OF_A_KIND
                            } else if (numJokers == 1) {
                                if (counts.keys.size == 3) {
                                    FULL_HOUSE
                                } else {
                                    THREE_OF_A_KIND
                                }
                            } else if (counts.keys.size == 3) {
                                TWO_PAIR
                            } else {
                                ONE_PAIR
                            }
                        1 -> when (numJokers) {
                            4 -> FIVE_OF_A_KIND
                            3 -> FOUR_OF_A_KIND
                            2 -> THREE_OF_A_KIND
                            1 -> ONE_PAIR
                            else -> HIGH_CARD
                        }
                        else -> FIVE_OF_A_KIND // All jokers case
                    }
                }
            }
        }
    }
}

File: swift/Solutions/Day07.swift
//
//  Day07.swift
//  advent_of_code_2023
//
//  Created by Kushantha Attanayake on 12/5/23.
//

import Foundation

struct Day07: Day {
    func part1Solution(input: [String]) -> String {
        var wagers = [(Hand, Int)]()
        input.forEach { round in
            let roundParts = round.split(separator: " ")
            wagers.append((Hand(cards: String(roundParts[0])), Int(roundParts[1])!))
        }
        wagers.sort {
            let (hand0, _) = $0
            let (hand1, _) = $1
            return hand0 < hand1
        }
        var answer = 0
        for (idx, element) in wagers.enumerated(){
            let (_, wage) = element
            answer += ((idx + 1) * wage)
        }
        return "\(answer)"
    }
    
    func part2Solution(input: [String]) -> String {
        var wagers = [(Hand, Int)]()
        input.forEach { round in
            let roundParts = round.split(separator: " ")
            wagers.append((Hand(cards: String(roundParts[0]), isPartTwo: true), Int(roundParts[1])!))
        }
        wagers.sort {
            let (hand0, _) = $0
            let (hand1, _) = $1
            return hand0 < hand1
        }
        var answer = 0
        for (idx, element) in wagers.enumerated(){
            let (_, wage) = element
            answer += ((idx + 1) * wage)
        }
        return "\(answer)"
    }
    
    
}

struct Hand: Hashable, Comparable {
    static func < (lhs: Hand, rhs: Hand) -> Bool {
        // Comparing by type first
        if(lhs.type != rhs.type){
            return lhs.type < rhs.type
        }
        
        // If of the same type, compare card by card
        for index in 0..<lhs.cards.count {
            if(lhs.cards[index] != rhs.cards[index]) {
                return lhs.cards[index] < rhs.cards[index]
            }
        }
        // If exactly identical, return falce
        return false
    }
    
    let cards: [CamelCard]
    let type: HandType
    
    init(cards: String) {
        self.cards = cards.map { CamelCard.fromChar($0) }
        var cardCounts = [Character:Int]()
        cards.forEach {
            cardCounts[$0] = cardCounts[$0, default: 0] + 1
        }
        let maxDuplicates = cardCounts.values.max()
        switch(maxDuplicates){
        case 5:
            type = HandType.five_of_a_kind
        case 4:
            type = HandType.four_of_a_kind
        case 3:
            if cardCounts.keys.count == 2 { type = HandType.full_house } else { type = HandType.three_of_a_kind }
        case 2:
            if cardCounts.keys.count == 3 { type = HandType.two_pair } else { type = HandType.one_pair }
        default:
            type = HandType.high_card
        }
    }
    
    init(cards: String, isPartTwo: Bool) {
        self.cards = cards.map { CamelCard.fromChar($0, isPartTwo: true) }
        var cardInts = [Character:Int]()
        var jokerCount = 0
        cards.forEach {
            if $0 == "J" {
                jokerCount += 1
            } else {
                cardInts[$0] = cardInts[$0, default: 0] + 1
            }
        }
        let maxDuplicates = cardInts.values.max()
        switch(maxDuplicates){
        case 5:
            type = HandType.five_of_a_kind
        case 4:
            if jokerCount == 1 {
                type = HandType.five_of_a_kind
            } else {
                type = HandType.four_of_a_kind
            }
        case 3:
            if (jokerCount == 2) {
                type = HandType.five_of_a_kind
            } else if (jokerCount == 1) {
                type = HandType.four_of_a_kind
            } else if cardInts.keys.count == 2 {
                type = HandType.full_house
            } else {
                type = HandType.three_of_a_kind
            }
        case 2:
            if jokerCount == 3 {
                type = HandType.five_of_a_kind
            } else if jokerCount == 2 {
                type = HandType.four_of_a_kind
            } else if jokerCount == 1 {
                if cardInts.keys.count == 2 {
                    type = HandType.full_house
                } else {
                    type = HandType.three_of_a_kind
                }
            } else if cardInts.keys.count == 3 {
                type = HandType.two_pair
            } else {
                type = HandType.one_pair
            }
        case 1:
            if jokerCount == 4 {
                type = HandType.five_of_a_kind
            } else if jokerCount == 3 {
                type = HandType.four_of_a_kind
            } else if jokerCount == 2 {
                type = HandType.three_of_a_kind
            } else if jokerCount == 1 {
                type = HandType.one_pair
            } else {
                type = HandType.high_card
            }
        default: // All joker case
            type = HandType.five_of_a_kind
        }
    }
    
}

enum HandType: Int, Comparable {
    static func < (lhs: HandType, rhs: HandType) -> Bool {
        lhs.rawValue > rhs.rawValue
    }
    
    case five_of_a_kind
    case four_of_a_kind
    case full_house
    case three_of_a_kind
    case two_pair
    case one_pair
    case high_card
}

enum CamelCard: Int, Comparable {
    static func < (lhs: CamelCard, rhs: CamelCard) -> Bool {
        lhs.rawValue > rhs.rawValue
    }
    
    case A, K, Q, J, T, NINE, EIGHT, SEVEN, SIX, FIVE, FOUR, THREE, TWO, JOKER
    
    static func fromChar(_ from: Character, isPartTwo:Bool = false) -> CamelCard {
        switch(from){
        case "A":
            return CamelCard.A
        case "K":
            return CamelCard.K
        case "Q":
            return CamelCard.Q
        case "J":
            if isPartTwo { return CamelCard.JOKER } else { return CamelCard.J }
        case "T":
            return CamelCard.T
        case "9":
            return CamelCard.NINE
        case "8":
            return CamelCard.EIGHT
        case "7":
            return CamelCard.SEVEN
        case "6":
            return CamelCard.SIX
        case "5":
            return CamelCard.FIVE
        case "4":
            return CamelCard.FOUR
        case "3":
            return CamelCard.THREE
        default:
            return CamelCard.TWO
        }
    }
}
